mlir/lib/Catalyst/Utils/CountStaticForloopIterations.cpp

```
@@ -0,0 +1,83 @@
+// Copyright 2025 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cmath> // std::ceil()
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Operation.h"
+
+using namespace mlir;
+
+namespace catalyst {
+
+static int64_t getNumIterations(int64_t lowerBound, int64_t upperBound, int64_t step)
+{
+    return std::ceil((upperBound - lowerBound) / step);
+}
+
+static int64_t getIntFromArithConstantOp(arith::ConstantOp op)
+{
+    // The magical incantation to get a cpp integer from an arith.constant op
+    assert(isa<IntegerAttr>(op.getValue()));
+    return cast<IntegerAttr>(op.getValue()).getValue().getSExtValue();
+}
+
+// Given an op in a for loop body with a static number of start, end and step,
+// compute the number of iterations that will be executed by the for loop.
+// Returns -1 if any of the above for loop information is not static.
+//
+// Note: if the input op is not inside any for loop operations,
+// this method returns 1, since there would be just one "iteration".
+int64_t countStaicForloopIterations(Operation *op)
+{
+    assert(!isa<scf::ForOp>(op));
```

Is this here to prevent nested for loops? 🤔  in that case maybe we should raise a meaningful error.

---

I see now that we only enter this function if visiting a `PPMeasurementOp` or `PPRotationOp` so not sure why we have the assert. Also just to confirm, we are supporting nested for loops right?

---

We do support nested for loops. I assert here because this is a utility, and having the counted op itself be a for loop op felt a bit weird to me.

I guess technically this assert is not needed, since "counting how many times a forloop op is occuring inside another for loop op" is a valid thing to do?